### ISSUE TITLE
Rebase apollo-server in middleware architecture

### ISSUE THREAD

**Issue by Rusfighter** [apollographql/apollo-server#1245]
Hey there,

Currently I am creating a simple apollo-server to query some graphql data. Since I want to go full serverless, I come across some design decisions that are currently made, that give me some problems.

In my case I am using firebase cloud functions to host apollo-server. The reason is because you get free CDN caching out of the box (see https://firebase.google.com/docs/hosting/functions , cache-control section). However since you can only use CDN caching with apollo engine, and I don't want to host any proxy servers, I reversed engineerd the apollo engine package and splitted up the apollo-server-express in several express middlewares where you can get much more control. 

The result will be something like this:

```
app.use(
  '/graphql',
  // checks the cache
  getFromCacheIfAny,
  // get response from graphql-resolvers if not in cache
  gExpress,
  // store in cache if it is a get/post query request
  storeInCache,
  // filter out the extensions
  extensionsFilter,
  // send the response
  graphqlResponseHandler
);
```

As you can see, you can plug multiple middleware to modify the request and response (concept of middleware). I think this design gives far more control to the developers to implement all sort of caching, logging, tracing, etc. 

A working example, with CDN caching & inmemory caching can be found here: https://github.com/Rusfighter/firegraph/tree/master/functions/src

**Comment by Rusfighter** [comment]
@evans Yes with this middleware design it is easy to implement full response caching. However I don't have time to create a PR for this since it will affect all the apollo-server-* packages.

**Comment by evans** [comment]
@Rusfighter The implementation will probably occur in apollo-server-core, so that all of the integrations would retain the same caching semantics. We'd probably be storing the result of `runQuery`
